env/export.go

```diff
@@ -5,8 +5,22 @@ import (
 	"strings"
 )
 
-var posixExportLineRegex = regexp.MustCompile("\\Adeclare \\-x ([a-zA-Z_]+[a-zA-Z0-9_]*)(=\")?(.+)?\\z")
-var endsWithUnescapedQuoteRegex = regexp.MustCompile("([^\\\\]\"\\z|\\A\"\\z)")
+var (
+	posixExportLineRegex = regexp.MustCompile("\\Adeclare \\-[aAfFgiIlnrtux]+ ([a-zA-Z_]+[a-zA-Z0-9_]*)(=\")?(.+)?\\z")
+	//                                                             ^
+	//                                                             |
+	//                                           all of the available options for declare
```

Comment:
Is it possible that being strict about matching a known set of options could break things in future if some version of bash or bash-emulating shell provides other (harmless) options?

Reply:
just coming back to this, because i'm now realising that i forgot to, post-merge: i think it's _probably_ fine - it's significantly more permissive now than it used to be - but IMO in the medium-term i think we need to move away from parsing the output of `export -p`, and move to parsing the output of `printenv` or something a bit less complex. I think if something comes along to further break this workflow, that should be the kick in the pants we need to properly refactor this.

A low-hanging fruit version of this might just be refactoring the logic of this parser from what we currently have:
```Go
firstLine := lines[0]
if looksLikeExportDashPOutput(firstLine) {
	for line := range lines {
		// Parse the entire file like it's export -p
	} 
} else {
  // It's _definitely_ "THIS=that" style output, parse it like that
}
```
to something like 
```Go
for line := range lines {
	k, v := ParseExportLine(line)
	env.Set(k, v)
}

func ParseExportLine(line string) (string, string, error) {
	switch {
		case LooksLikeExportDashP(line):
			return ParseExportDashP(line)
		case LooksLikeEnv(line):
			return ParseEnv(line)
		default:
			fmt.Println("Failed to parse line:", line)
			return "", "", fmt.Errorf("Failed to parse line: %v", line)
	}
	return parts[0], parts[1]
}
```

that is, currently, anything that fails to get parsed using our magic export -p regex is assumed to be env-style (eg `THIS=that`), so any failure cases fall through the cracks. We should probably update this parser to be aware of error cases in general, which it's currently not.

this is all a problem for future benno though - at the moment, i think what we have is acceptable, but i agree that it could be better